lib/Sema/ConstantnessSemaDiagnostics.cpp

```diff
@@ -335,25 +335,56 @@ static void diagnoseConstantArgumentRequirementOfCall(const CallExpr *callExpr,
 void swift::diagnoseConstantArgumentRequirement(
     const Expr *expr, const DeclContext *declContext) {
   class ConstantReqCallWalker : public ASTWalker {
-    const ASTContext &astContext;
+    DeclContext *DC;
 
   public:
-    ConstantReqCallWalker(ASTContext &ctx) : astContext(ctx) {}
+    ConstantReqCallWalker(DeclContext *DC) : DC(DC) {}
 
     // Descend until we find a call expressions. Note that the input expression
     // could be an assign expression or another expression that contains the
     // call.
     std::pair<bool, Expr *> walkToExprPre(Expr *expr) override {
-      if (!expr || isa<ErrorExpr>(expr) || !expr->getType())
+      // Handle closure expressions separately as we may need to
+      // manually descend into the body.
+      if (auto *closureExpr = dyn_cast<ClosureExpr>(expr)) {
+        return walkToClosureExprPre(closureExpr);
+      }
+      // Interpolated expressions' bodies will be type checked
+      // separately so exit early to avoid duplicate diagnostics.
+      if (!expr || isa<ErrorExpr>(expr) || !expr->getType() ||
+          isa<InterpolatedStringLiteralExpr>(expr))
```

Comment:
Through adding new tests, I found that the original source of duplicate diagnostics has to do with the fact that string interpolations are getting type-checked separately for the expression of the type that conforms to ExpressibleByStringInterpolation and the expression of the actual interpolations. This causes a complication because we were expecting that we need to manually check inside of single-expression closures, but that is not the case when the single expression is one of these string interpolations.

Reply:
To remedy this, I'm skipping `InterpolatedStringLiteralExpr` types here, because they're getting checked when we walk the statements via `walkToStmtPre`.

Reply:
Makes sense! I keep working that interpolations use a special type-check entry point...